runtime/near-vm-runner/src/errors.rs

```diff
@@ -58,33 +64,35 @@ impl IntoVMError for wasmer_runtime::error::ResolveError {
 
 impl IntoVMError for wasmer_runtime::error::RuntimeError {
     fn into_vm_error(self) -> VMError {
-        use wasmer_runtime::error::RuntimeError;
-        match &self {
-            RuntimeError::Trap { msg } => {
-                VMError::FunctionCallError(FunctionCallError::WasmTrap { msg: msg.to_string() })
-            }
-            RuntimeError::Error { data } => {
-                if let Some(err) = data.downcast_ref::<VMLogicError>() {
-                    match err {
-                        VMLogicError::HostError(h) => {
-                            VMError::FunctionCallError(FunctionCallError::HostError(h.clone()))
-                        }
-                        VMLogicError::ExternalError(s) => VMError::ExternalError(s.clone()),
-                        VMLogicError::InconsistentStateError(e) => {
-                            VMError::InconsistentStateError(e.clone())
-                        }
-                    }
-                } else {
-                    eprintln!(
-                        "Bad error case! Output is non-deterministic {:?} {:?}",
-                        data.type_id(),
-                        self.to_string()
-                    );
-                    VMError::FunctionCallError(FunctionCallError::WasmTrap {
-                        msg: "unknown".to_string(),
-                    })
+        use wasmer_runtime::ExceptionCode;
+        let data = &*self.0;
+
+        // TODO: Wasmer provides no way to distingush runtime Internal Wasmer errors or host panics
+        // (at least for a single-pass backend)
+        // https://github.com/wasmerio/wasmer/issues/1338
+        if let Some(err) = data.downcast_ref::<VMLogicError>() {
+            match err {
+                VMLogicError::HostError(h) => {
+                    VMError::FunctionCallError(FunctionCallError::HostError(h.clone()))
+                }
+                VMLogicError::ExternalError(s) => VMError::ExternalError(s.clone()),
+                VMLogicError::InconsistentStateError(e) => {
+                    VMError::InconsistentStateError(e.clone())
                 }
             }
+        } else if let Some(_) = data.downcast_ref::<ExceptionCode>() {
+            VMError::FunctionCallError(FunctionCallError::WasmTrap { msg: format!("{}", self) })
+        } else if let Some(_) = data.downcast_ref::<String>() {
+            VMError::FunctionCallError(FunctionCallError::WasmTrap { msg: format!("{}", self) })
+        } else if let Some(_) = data.downcast_ref::<&str>() {
+            VMError::FunctionCallError(FunctionCallError::WasmTrap { msg: format!("{}", self) })
+        } else {
```

Comment:
wait why?
are these guaranteed to be deterministic?

Reply:
We don't know yet. I'll add warning also.

Reply:
The reason we returned `"unknown"` is that at least in earlier wasmer it could return a string with contents that differ between nodes or between two runs on one node.

If we return the string as result of transaction execution, we need to be sure that an execution with the same input always produces the same string.